mono/metadata/threads.c

```diff
@@ -3603,6 +3603,97 @@ get_thread_dump (MonoThreadInfo *info, gpointer ud)
 	return MonoResumeThread;
 }
 
+typedef struct QuickDumpUserData {
+	char *bufp;
+	char *maxp;
+	const MonoUnityCallstackOptions *opts;
+} QuickDumpUserData;
+
+static void
+append_quick (QuickDumpUserData *ud, const char *str)
+{
+	int len = (int)strlen (str);
+	int remain = (int)(ud->maxp - ud->bufp);
+	int copylen = len;
+	if (copylen > remain)
+		copylen = remain;
+	memcpy (ud->bufp, str, copylen);
+	ud->bufp += copylen;
+}
+
+static int
+append_parent_types (QuickDumpUserData *ud, const MonoClass *k)
+{
+	if (!k)
+		return 0;
+
+	if (append_parent_types (ud, k->parent)) {
+		append_quick (ud, ".");
+	}
+
+	append_quick (ud, k->name);
+	return 1;
+}
+
+static mono_bool
+collect_frame_text (MonoMethod *method, int32_t native_offset, int32_t il_offset, mono_bool managed, void *data)
+{
+	QuickDumpUserData *ud = (QuickDumpUserData *)data;
+
+	if (managed && method) {
+		char *method_name = mono_method_full_name (method, TRUE);
```

Comment:
Note to myself, we should provide a no-alloc or temp alloc version of this and similar functions.